### PR TITLE
[v1.0.0] Flexible form controls, ExactOut SwapMode , revamped app preview

### PR DESCRIPTION
**Pull request by Arrowana**
I think it is affected by the ordering of quote being incorrect in v3 for ExactOut, do we want to backport the recent changes? Otherwise exactout of v3 is a bit of a death trap.

what is left
- [ ] make the ui a bit more explicit as to which input is editable
- [ ] can we convey to the user if it is in exact output mode?
- [ ] should user be able to toggle swap mode? and integrator then disable or not that option
- [x] add some buttons/checkboxes in the demo so the integrator can test it out
- [x] ~~glitchy on first load as we have prefilled some values but tokenMap arrives only later and we default decimals to 0 (rug)~~ fixed, no tokenMap, no prefill
- [ ] v3 fixes in the api, to fix quotes in random order (ordered in api by outAmount, equal for all in exact output mode)
- [ ] selection of route shows outAmount, so the same for all routes

### CODE REVIEW

File: src/contexts/SwapContext.tsx [hunk 1]
```diff
@@ -149,14 +173,36 @@ export const SwapContextProvider: FC<{
...
-    if (!form.fromValue || !fromTokenInfo) return JSBI.BigInt(0);
+  // Set value given initial amount
+  useEffect(() => {
+    if (!initialAmount) return;
+
+    const toUiAmount = (mint: string) => {
+      const tokenInfo = mint ? tokenMap.get(mint) : undefined;
+      if (!tokenInfo) return;
+      return String(fromLamports(JSBI.BigInt(initialAmount), tokenInfo.decimals));
+    };
```
Comment (dicksonpys) [inline src/contexts/SwapContext.tsx:184]: can move into utils

**Review by dicksonpys — COMMENTED** [review]
adding some comments, I will help fix

File: src/components/Form.tsx [hunk 2]
```diff
@@ -91,17 +104,28 @@ const Form: React.FC<{
...
-        toMint: prev.fromMint,
-      }));
-    }
+    setForm((prev) => ({
+      ...prev,
+      fromValue: '',
+      toValue: '',
+      fromMint: prev.toMint,
+      toMint: prev.fromMint,
+    }));
```
[the remark this answers is not in the document]
Comment (Arrowana) [reply src/components/Form.tsx:113]: I want to remove outputOnly for the more flexible configure everything from init

File: src/components/Form.tsx [hunk 3]
```diff
@@ -91,17 +104,28 @@ const Form: React.FC<{
...
+    const result = { inputAmountDisabled: true, outputAmountDisabled: true };
+    if (!fixedAmount) {
+      if (swapMode === SwapMode.ExactOut) {
+        result.outputAmountDisabled = false;
+      } else {
+        result.inputAmountDisabled = false;
+      }
+    }
+    return result;
+  }, [fixedAmount, swapMode]);
```
[the remark this answers is not in the document]
Comment (dicksonpys) [reply src/components/Form.tsx:127]: looks good, ps.

File: package.json [hunk 4]
```diff
@@ -29,13 +29,16 @@
     "@solana/wallet-adapter-wallets": "0.19.10",
     "@solana/web3.js": "1.73.0",
     "@svgr/webpack": "^6.5.0",
+    "@types/lodash.debounce": "^4.0.7",
+    "@types/react-syntax-highlighter": "^15.5.6",
```
Comment (ZhengYuTay) [inline package.json:33]: let's move some things to devDependencies?

- @types/*
- build tools (autoprefixer)
- tailwindcss
- stylus
- @svgr/webpack

File: src/components/screens/InitialScreen.tsx [hunk 5]
```diff
@@ -27,37 +25,38 @@ const InitialScreen = ({ mint, setIsWalletModalOpen, isWalletModalOpen }: Props)
...
-  }, [mint]);
-
-  // TODO: Dedupe the balance
   const balance = useMemo(() => {
     return form.fromMint ? accounts[form.fromMint]?.balance || 0 : 0;
   }, [walletPublicKey, accounts, form.fromMint]);
 
-  const isDisabled = useMemo(() => {
+  const [isDisabled, setIsDisabled] = useState(false);
+  useEffect(() => {
```
Comment (ZhengYuTay) [inline src/components/screens/InitialScreen.tsx:43]: not sure why the change to state

Comment (dicksonpys) [reply src/components/screens/InitialScreen.tsx:43]: had some console errors during mounting
did not understand the full extent of it, but might be related to useMemo being called before render, and setErrors called too early and the component wasn’t mounted properly yet

File: package.json (hunk 4 above)
Comment (dicksonpys) [reply package.json:33]: will do

**Review by ZhengYuTay — APPROVED** [review]

**Review by ZhengYuTay — APPROVED** [review]
